examples/binaries/calc-hash/over-blocks/src/code.rs

```diff
@@ -0,0 +1,71 @@
+use crate::Method;
+use gstd::{exec, msg, ActorId, BTreeMap, Decode, Encode, MessageId, Vec};
+use shared::{Package, PackageId};
+
+mod state {
+    use gstd::{ActorId, BTreeMap, Vec};
+    use shared::{Package, PackageId};
+
+    pub static mut THRESHOLD: Option<u64> = None;
+    pub static mut REGISTRY: BTreeMap<PackageId, Package> = BTreeMap::new();
+}
+
+#[no_mangle]
+unsafe extern "C" fn init() {
+    state::THRESHOLD = Some(msg::load().expect("Invalid threshold."));
+}
+
+#[gstd::async_main]
+async fn main() {
+    let method = msg::load::<Method>().expect("Invalid contract method.");
+
+    unsafe {
+        let threshold = state::THRESHOLD.expect("Threshold has not been set.");
+
+        match method {
+            Method::Start(pkg_with_id) => {
+                state::REGISTRY.insert(pkg_with_id.id, pkg_with_id.package);
+            }
+            Method::Refuel(id) => dispatch(id).await,
+            Method::Calculate(mut pkg) => {
+                while exec::gas_available() > threshold {
+                    pkg = pkg.calc();
+
+                    if pkg.finished() {
+                        break;
+                    }
+                }
+
+                let _ = msg::reply(pkg, 0).expect("Send reply failed.");
+            }
+        }
```

Comment:
(always remember that the algorithmic language is our everything xD)
this stuff in words should look like this:

STORING DATA:
In memory, you keep mapping PackageId -> (MessageId, Package), where message id is initial one, which started calculations.

INPUT and logic:
Method::Start:
```rust
if !registry.contains(id) {
    registry.insert(id, (msg_id, pkg))
}

let (_, pkg_re)f = registry.get(id);

if pkg_ref.finished() {
    msg::reply(pkg_ref.result())
} else {
    wait()
}
```

Method::Refuel:
```rust
let (&message_id, pkg_ref) = registry.get(id).expect("Wrong id");
while can_calculate() {
   pkg_ref.calc();
   
   if pkg_ref.finished() {
       wake(message_id);
       return
   }
}

//optionally
msg::reply("user, give me more gas");
```

Method::Calculate seems excess in such arch for me

Reply:
> Method::Start:
>
> ```rust
> if !registry.contains(id) {
>     registry.insert(id, (msg_id, pkg))
> }
> 
> let (_, pkg_re)f = registry.get(id);
> 
> if pkg_ref.finished() {
>     msg::reply(pkg_ref.result())
> } else {
>     wait()
> }
> ```

As I saw in logs, the logic before `wait` will be executed before wake, for the logic of this snippet, do you mean the logic of wait will execute for twice? one before `wait` one after `wake`?

Reply:
> > Method::Start:
> > ```rust
> > if !registry.contains(id) {
> >     registry.insert(id, (msg_id, pkg))
> > }
> > 
> > let (_, pkg_re)f = registry.get(id);
> > 
> > if pkg_ref.finished() {
> >     msg::reply(pkg_ref.result())
> > } else {
> >     wait()
> > }
> > ```
> 
> As I saw in logs, the logic before `wait` will be executed before wake, for the logic of this snippet, do you mean the logic before wait will execute for twice? one before `wait` one after `wake`?

yeah, because there is no async context. with such code whole demo may be sync

Reply:
if pre-wait logic will be executed after wake, that won't break anything